src/compiler/checker.ts

```diff
@@ -19643,10 +19644,17 @@ namespace ts {
             return declaredType;
         }
 
-        function getTypeFactsOfTypes(types: Type[]): TypeFacts {
+        function getTypeFactsOfTypes(types: Type[], isUnion: boolean): TypeFacts {
             let result: TypeFacts = TypeFacts.None;
             for (const t of types) {
-                result |= getTypeFacts(t);
+                const facts = getTypeFacts(t);
+                result |= facts;
+            }
+            if (!isUnion) {
+                // Get the set of postive facts for the intersection by masking with the negative set shifted left, then shift those present positive facts into the negative fact
+                // value range, and unset any of those bits (by negating that mask and then intersecting it with the original value)
+                const positiveFacts = result & (TypeFacts.NegativeTypeofFacts << 7);

```

Comment:
You are _absolutely_ right - the core to fixing the reported bug was excluding the negative facts from the result, however during one of the refactorings the `>>` and the `<<` on the next line got swapped, so we weren't keeping _any_ negative fact flags (plus an off by one error because I'm bad at counting).